projects/ssr-tests/src/ssr-testing.spec.ts

```diff
@@ -33,15 +33,18 @@ describe('SSR E2E', () => {
     ]);
   });
 
-  // TODO: Test incomplete
-  xit('should receive cached response with next request', async () => {
+  it('should receive cached response with next request', async () => {
+    await Ssr.startSsrServer();
     proxy = await ProxyServer.startProxyServer({
       target: BACKEND_BASE_URL,
     });
     const response: any = await ProxyServer.sendRequest(REQUEST_PATH);
     expect(response.statusCode).toEqual(200);
+    Log.waitUntilLogContainsText(`Rendering completed (${REQUEST_PATH})`);
 
-    Log.assertMessages(['Render from cache (/)']);
+    const response2: any = await ProxyServer.sendRequest(REQUEST_PATH);
+    expect(response2.statusCode).toEqual(200);
+    Log.assertMessages([`Render from cache (${REQUEST_PATH})`]);
```

Comment:
[QUESTION]
How do we ensure that the previous request _really_ timed out, so the next request _for sure_ will get response from the cache?

Reply:
I thought that the SSR engine retains the request in cache regardless of timeout so that it could serve the cached page to future requests for that url (ie. render page first time, serve cached in future). Is there another purpose for this feature? I may have understood the mechanism.

Reply:
Short answer:
In general Spartacus by default doesn't cache rendered results for future use (unless the `cache: true` SsrOptimizationOption is set, which is not recommended - CDN caching should be used instead).

_However_ Spartacus by default caches the result in one exceptional case, and only for a _single_ usage in the future - when the original request timed out but the render is still pending in the background. Then we don't waste this eventually rendered result - we put it into a cache for a single use. Then we return it immediately to the very next incoming request. And we delete it from the cache. See the [spec](https://github.com/SAP/spartacus/blob/b59e27b5011fd79096731e94c1f58cc4a34e8ec3/core-libs/setup/ssr/optimized-engine/optimized-ssr-engine.spec.ts#L182-L190).

---
Long answer:
Here is a step-by-step explanation of how SAP Spartacus SSR (Server-Side Rendering) works in the case of fallback to CSR (Client-Side Rendering) when the request times out due to the rendering taking too long. (I wrote it in the past somewhere, now I'm just pasting it here for knowledge sharing):

1. When an incoming request comes to the NodeJS server, the server starts counting the time for the request timeout. This timeout is typically set to a specific duration, after which the server will return a response to the client.

2. The rendering process starts, which involves generating the HTML content for the requested page. However, if the rendering process takes longer than the request timeout duration, the server will not be able to wait for the rendering to complete before returning a response.

3. In the meantime, as the request timeout time passes, the server decides to return a CSR (Client-Side Rendering) response to the request. This means that instead of waiting for the server-side rendering to complete, the server sends back a response that includes the necessary JavaScript files and instructions for the client to perform the rendering on the browser side.

4. It's important to note that even though the server has returned a CSR response, the rendering process is still pending in the background. This is because in Angular SSR, there is no way to abort a pending rendering.

5. Eventually, the rendering process completes, and the server logs a message indicating that the rendering has been completed. However, since the response has already been sent to the client, this completion message is logged to the server's console and does not affect the response that has already been returned.

The reason for continuing the rendering process in the background even though the request has already been answered is because in Spartacus SSR, there are additional features that take advantage of this behavior.

One such feature is the "reuseCurrentRendering" functionality. If another request comes in for the same URL while the rendering is still pending, and the rendering completes within the new request's timeout window, then this new request "wins" the response of the completed rendering. This means that the client will receive the fully rendered page as a response.

If there are multiple pending requests for the same URL at the moment when the rendering completes, all of them will "win" the response of the completed rendering.

If no request was awaiting the rendering completion for a specific URL, then the result HTML is stored in the in-memory cache, waiting for the next request. The very next request for the same URL will get an immediate response taken from the cache, and the cached HTML will be cleared. However, if you have configured the cache to be kept longer (using the `cache:true` configuration), the cache will be retained for a longer period, although this is not recommended.

I hope this explanation helps! Let me know if you have any further questions.

Reply:
Thanks for the thorough explanation! I wonder if there is a need to test the rendering caching again in these e2es if it is already tested in the link you provided?